File: packages/zynd-shared/src/notifications/deep-links.ts
export type NotificationCategory = "security" | "kyc" | "referral" | "account" | "family";

export type NotificationDeepLinkInput = {
  notification_type: string;
  category?: NotificationCategory;
  metadata?: Record<string, unknown> | null;
};

export type NotificationDeepLink = {
  path: string;
  settingsSection?: string;
};

const SECURITY_SETTINGS_SECTION = "security";
const DEVICES_SETTINGS_SECTION = "your-devices";
const NOTIFICATIONS_SETTINGS_SECTION = "notifications";

const NOTIFICATION_TYPE_ROUTES: Record<string, NotificationDeepLink> = {
  "auth.login.succeeded": { path: "/dashboard/settings", settingsSection: SECURITY_SETTINGS_SECTION },
  "auth.login.failed": { path: "/dashboard/settings", settingsSection: SECURITY_SETTINGS_SECTION },
  "auth.new_device": { path: "/dashboard/settings", settingsSection: DEVICES_SETTINGS_SECTION },
  "auth.password.changed": { path: "/dashboard/settings", settingsSection: SECURITY_SETTINGS_SECTION },
  "auth.email.changed": { path: "/dashboard/settings", settingsSection: "email" },
  "auth.email_change.requested": { path: "/dashboard/settings", settingsSection: "email" },
  "auth.mfa.enabled": { path: "/dashboard/settings", settingsSection: SECURITY_SETTINGS_SECTION },
  "auth.mfa.disabled": { path: "/dashboard/settings", settingsSection: SECURITY_SETTINGS_SECTION },
  "auth.device.revoked": { path: "/dashboard/settings", settingsSection: DEVICES_SETTINGS_SECTION },
  "auth.refresh_reuse.detected": { path: "/dashboard/settings", settingsSection: SECURITY_SETTINGS_SECTION },
  "auth.pin.set": { path: "/dashboard/settings", settingsSection: SECURITY_SETTINGS_SECTION },
  "auth.pin.reset": { path: "/dashboard/settings", settingsSection: SECURITY_SETTINGS_SECTION },
  "kyc.initiated": { path: "/dashboard/kyc" },
  "kyc.under_review": { path: "/dashboard/kyc" },
  "kyc.completed": { path: "/dashboard/kyc" },
  "kyc.rejected": { path: "/dashboard/kyc" },
  "referral.user.signed_up": { path: "/dashboard/referral/referrals" },
  "referral.kyc.verified": { path: "/dashboard/referral/referrals" },
  "referral.first_investment": { path: "/dashboard/referral/referrals" },
  "referral.qualified": { path: "/dashboard/referral" },
  "referral.engaged": { path: "/dashboard/referral" },
  "account.settings.changed": {
    path: "/dashboard/settings",
    settingsSection: NOTIFICATIONS_SETTINGS_SECTION,
  },
  "account.profile_image.updated": {
    path: "/dashboard/settings",
    settingsSection: "personal-details",
  },
  "invest.risk_profile.completed": {
    path: "/dashboard/risk-profile",
  },
  "invest.risk_profile.unlock_otp": {
    path: "/dashboard/notifications",
  },
  "family.invite.received": {
    path: "/dashboard/family",
  },
  "family.invite.accepted": {
    path: "/dashboard/family",
  },
  "family.invite.declined": {
    path: "/dashboard/family",
  },
  "family.member.removed": {
    path: "/dashboard/family",
  },
  "family.member.role_changed": {
    path: "/dashboard/family",
  },
  "family.head.transferred": {
    path: "/dashboard/family",
  },
  "family.invite.reminder": {
    path: "/dashboard/family",
  },
};

const CATEGORY_FALLBACK_ROUTES: Record<NotificationCategory, NotificationDeepLink> = {
  security: { path: "/dashboard/settings", settingsSection: SECURITY_SETTINGS_SECTION },
  kyc: { path: "/dashboard/kyc" },
  referral: { path: "/dashboard/referral" },
  account: { path: "/dashboard/settings", settingsSection: NOTIFICATIONS_SETTINGS_SECTION },
  family: { path: "/dashboard/family" },
};

export function resolveNotificationDeepLink(
  input: NotificationDeepLinkInput,
): NotificationDeepLink {
  const explicit = NOTIFICATION_TYPE_ROUTES[input.notification_type];
  if (explicit) {
    return explicit;
  }

  if (input.category) {
    return CATEGORY_FALLBACK_ROUTES[input.category];
  }

  return { path: "/dashboard/notifications" };
}

export function buildNotificationDeepLinkUrl(
  baseUrl: string,
  input: NotificationDeepLinkInput & { notification_id?: string },
): string {
  const resolved = resolveNotificationDeepLink(input);
  const url = new URL(resolved.path, baseUrl.endsWith("/") ? baseUrl : `${baseUrl}/`);

  if (resolved.settingsSection) {
    url.searchParams.set("section", resolved.settingsSection);
  }

  if (input.notification_id) {
    url.searchParams.set("notification_id", input.notification_id);
  }

  url.searchParams.set("notification_type", input.notification_type);
  return `${url.pathname}${url.search}`;
}

export const PUSH_DEVICE_STORAGE_KEY = "zynd_push_device_id";

File: Backend/app/application/notifications/deep_links.py
from __future__ import annotations

from typing import Any

from app.infrastructure.persistence.notification_models import NotificationCategory

SECURITY_SETTINGS_SECTION = "security"
NOTIFICATIONS_SETTINGS_SECTION = "notifications"

NOTIFICATION_TYPE_ROUTES: dict[str, tuple[str, str | None]] = {
    "auth.login.succeeded": ("/dashboard/settings", SECURITY_SETTINGS_SECTION),
    "auth.login.failed": ("/dashboard/settings", SECURITY_SETTINGS_SECTION),
    "auth.new_device": ("/dashboard/settings", SECURITY_SETTINGS_SECTION),
    "auth.password.changed": ("/dashboard/settings", SECURITY_SETTINGS_SECTION),
    "auth.email.changed": ("/dashboard/settings", "email"),
    "auth.email_change.requested": ("/dashboard/settings", "email"),
    "auth.mfa.enabled": ("/dashboard/settings", "mfa"),
    "auth.mfa.disabled": ("/dashboard/settings", "mfa"),
    "auth.device.revoked": ("/dashboard/settings", SECURITY_SETTINGS_SECTION),
    "auth.refresh_reuse.detected": ("/dashboard/settings", SECURITY_SETTINGS_SECTION),
    "auth.pin.set": ("/dashboard/settings", SECURITY_SETTINGS_SECTION),
    "auth.pin.reset": ("/dashboard/settings", SECURITY_SETTINGS_SECTION),
    "kyc.initiated": ("/dashboard/kyc", None),
    "kyc.under_review": ("/dashboard/kyc", None),
    "kyc.completed": ("/dashboard/kyc", None),
    "kyc.rejected": ("/dashboard/kyc", None),
    "referral.user.signed_up": ("/dashboard/referral/referrals", None),
    "referral.kyc.verified": ("/dashboard/referral/referrals", None),
    "referral.first_investment": ("/dashboard/referral/referrals", None),
    "referral.qualified": ("/dashboard/referral", None),
    "referral.engaged": ("/dashboard/referral", None),
    "account.settings.changed": ("/dashboard/settings", NOTIFICATIONS_SETTINGS_SECTION),
    "account.profile_image.updated": ("/dashboard/settings", "personal-details"),
    "invest.risk_profile.completed": ("/dashboard/risk-profile", None),
    "invest.risk_profile.unlock_otp": ("/dashboard/notifications", None),
    "invest.mitra_txn_recommendation": ("/dashboard/mutual-funds/recommendation", None),
    "family.invite.received": ("/dashboard/family", None),
    "family.invite.accepted": ("/dashboard/family", None),
    "family.invite.declined": ("/dashboard/family", None),
    "family.member.removed": ("/dashboard/family", None),
    "family.member.role_changed": ("/dashboard/family", None),
    "family.head.transferred": ("/dashboard/family", None),
    "family.invite.reminder": ("/dashboard/family", None),
}

CATEGORY_FALLBACK_ROUTES: dict[NotificationCategory, tuple[str, str | None]] = {
    NotificationCategory.security: ("/dashboard/settings", SECURITY_SETTINGS_SECTION),
    NotificationCategory.kyc: ("/dashboard/kyc", None),
    NotificationCategory.referral: ("/dashboard/referral", None),
    NotificationCategory.account: ("/dashboard/settings", NOTIFICATIONS_SETTINGS_SECTION),
    NotificationCategory.family: ("/dashboard/family", None),
}


def resolve_notification_deep_link_path(
    *,
    notification_type: str,
    category: NotificationCategory,
) -> tuple[str, str | None]:
    explicit = NOTIFICATION_TYPE_ROUTES.get(notification_type)
    if explicit:
        return explicit
    return CATEGORY_FALLBACK_ROUTES.get(category, ("/dashboard/notifications", None))


def build_notification_web_url(
    *,
    frontend_url: str,
    notification_id: str,
    notification_type: str,
    category: NotificationCategory,
    metadata: dict[str, Any] | None = None,
) -> str:
    recommendation_token = (metadata or {}).get("recommendation_token")
    if notification_type == "invest.mitra_txn_recommendation" and recommendation_token:
        base = frontend_url.rstrip("/")
        return (
            f"{base}/dashboard/mutual-funds/recommendation/{recommendation_token}"
            f"?notification_id={notification_id}&notification_type={notification_type}"
        )

    path, settings_section = resolve_notification_deep_link_path(
        notification_type=notification_type,
        category=category,
    )
    base = frontend_url.rstrip("/")
    query_parts = [f"notification_id={notification_id}", f"notification_type={notification_type}"]
    if settings_section:
        query_parts.append(f"section={settings_section}")
    return f"{base}{path}?{'&'.join(query_parts)}"
